### DISCUSSION TITLE
Go 语言反射的实现原理 · /golang-reflect

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#569] [Announcements]
https://draveness.me/golang-reflect

**Comment by amtoor** [comment]
**3.1**节里有个笔误：

> 通过 `emptyInterface` 这个**接口体**来表示的

应该是：通过 `emptyInterface` 这个结构体来表示的

**Comment by draveness** [comment]
> 应该是：通过 emptyInterface 这个结构体来表示的

我改一下

**Comment by draveness** [comment]
> 666

777

**Comment by tidyoux** [comment]
在 4.3.4 的结尾处似乎有些表述不清：

> 在其他情况下，由于方法都是按照字母序存储的，reflect.implements 会维护两个用于遍历接口和类型方法的索引 i 和 j 判断类型是否实现了接口，因为最多只会进行 n + m 次数的比较，所以整个过程的时间复杂度是 O(n+m)。

这里的 n 和 m 代表的内容没有说明，实际上按照原代码，for 循环最多执行次数 = 类型 V 的方法数，与接口 T 的方法数无关。

另外，这个 `implements` 函数似乎可以优化一下，通过判断 **V 方法数 < T 方法数** 而提前返回。

**Reply by unifreak** [reply]
我觉得这里还是应该是 O(n+m), 原代码基本上可以简化为下面, 本质上是对两个有序集合判断是否属于被属于的关系:

```go
        i := 0
		for j := 0; j < len(v.Methods); j++ {
            if inset() {
				if i++; i >= len(t.Methods) {
					return true
				}
			}
		}
```

一个算是官方的佐证 (https://research.swtch.com/interfaces):

> In general there might be ni methods for the interface type and nt methods for the concrete type. The obvious search to find the mapping from interface methods to concrete methods would take O(ni × nt) time, but we can do better. By sorting the two method tables and walking them simultaneously, we can build the mapping in O(ni + nt) time instead.

**Comment by draveness** [comment]
> @tidyoux
> 在 4.3.4 的结尾处似乎有些表述不清：
> 
> > 在其他情况下，由于方法都是按照字母序存储的，reflect.implements 会维护两个用于遍历接口和类型方法的索引 i 和 j 判断类型是否实现了接口，因为最多只会进行 n + m 次数的比较，所以整个过程的时间复杂度是 O(n+m)。
> 
> 这里的 n 和 m 代表的内容没有说明，实际上按照原代码，for 循环最多执行次数 = 类型 V 的方法数，与接口 T 的方法数无关。

修复了

> 另外，这个 `implements` 函数似乎可以优化一下，通过判断 **V 方法数 < T 方法数** 而提前返回。

你可以直接提个 PR，哈哈哈

**Comment by xue8** [comment]
111

**Comment by hoset** [comment]
只看了反射一章 。这应该是目前看到的，讲得最透彻的。

**Comment by hevervie** [comment]
硬核，写的太好了

**Comment by muyikouzi** [comment]
文章写得都很棒，很细节深入啊 收藏了！

**Comment by bambuo** [comment]
出版后公众号通知一下,一定买本好好收藏

**Comment by draveness** [comment]
> 出版后公众号通知一下,一定买本好好收藏

哈哈哈，好的

**Comment by KylinMountain** [comment]
有1个问题请教一下大神
1. 上一节的eface和这一节的emptyInterface结构体，是同一个东西吧？为啥表述以及内部字段类型不同了呢？虽然表示的都是同一个东西，类型和数据。是否要统一表述呢？

然后，看过上一节接口之后，了解了`interface{}`的eface之后，理解这里的reflect.ValueOf(i interface{})和reflect.TypeOf(i interface{})，以及它们的双向类型转换，感觉毫无障碍。

另外，我感觉调用`implements(u.(*rtype), t)` 和 方法 `func implements(T, V *rtype) bool {}`参数值命名有种蛋疼的感觉，调用时入参还是第二个是t，到方法声明上时，已经变成了第一个是T了。。。虽然一个是大写，一个小写，还是容易让人错乱一下。

最后，方法调用，就是函数调用那一节的入栈出栈那一套嘛，只是由reflect自己去读取解析函数入参出参，以及方法执行啥的。

还是大神讲的透彻，出书时，我也要买一本~就是实体书还能呈现出这么色彩丰富的图吗？

**Comment by draveness** [comment]
> @KylinGu 有1个问题请教一下大神
> 
> 1. 上一节的eface和这一节的emptyInterface结构体，是同一个东西吧？为啥表述以及内部字段类型不同了呢？虽然表示的都是同一个东西，类型和数据。是否要统一表述呢？

观察的非常细啊，这两个结构体中的前者主要用于 Go 语言运行时，后者用于反射。

> 还是大神讲的透彻，出书时，我也要买一本~就是实体书还能呈现出这么色彩丰富的图吗？

这个不好说，双色和彩色的成本会差很多，最后也会反应在价格上

**Comment by KylinMountain** [comment]
> @tidyoux 
> 这里的 n 和 m 代表的内容没有说明，实际上按照原代码，for 循环最多执行次数 = 类型 V 的方法数，与接口 T 的方法数无关。
> 
> 另外，这个 `implements` 函数似乎可以优化一下，通过判断 **V 方法数 < T 方法数** 而提前返回。

我觉得如果implements 是rtype的方法的话，可以考虑通过判断rtype的方法数量小于入参的方法数量提前退出。但是它是函数，不能假定谁是接口实现方，谁是接口提供方。
`func implements(T, V *rtype) bool {
...
}`

**Comment by shuff1e** [comment]
> @draveness
> > @KylinGu 有1个问题请教一下大神
> > 
> > 1. 上一节的eface和这一节的emptyInterface结构体，是同一个东西吧？为啥表述以及内部字段类型不同了呢？虽然表示的都是同一个东西，类型和数据。是否要统一表述呢？
> 
> 观察的非常细啊，这两个结构体中的前者主要用于 Go 语言运行时，后者用于反射。
> 
> > 还是大神讲的透彻，出书时，我也要买一本~就是实体书还能呈现出这么色彩丰富的图吗？
> 
> 这个不好说，双色和彩色的成本会差很多，最后也会反应在价格上
> 

不如直接上线到极客时间上，这样卖钱更快一些

**Comment by bspp1314** [comment]
楼主，请教一个问题，定义了一下的数据类型

type Status int
type RealStatus Status 

有没有办法通过反射从 RealStatus Type 获取到 Status 的Type

**Comment by draveness** [comment]
> 楼主，请教一个问题，定义了一下的数据类型
> 
> type Status int
> type RealStatus Status
> 
> 有没有办法通过反射从 RealStatus Type 获取到 Status 的Type

根据这个 Proposal 可能确实不太行 https://github.com/golang/proposal/blob/master/design/18130-type-alias.md#reflect 但是也不排除它新支持了其他方法

**Comment by wuqinqiang** [comment]
没明白 为什么需要调用 reflect.escapes 保证当前值逃逸到堆上，如果在栈上会则样？

**Comment by Northward712** [comment]
楼主您好 ，如果总结以下reflect是不是可以理解为就是通过利用interface的runtime中eface转换为反射中的emtpyInterface字段以取出其type字段和data字段去对其原本的数据和类型进行一些操作

**Comment by Northward712** [comment]
> @wuqinqiang
> 没明白 为什么需要调用 reflect.escapes 保证当前值逃逸到堆上，如果在栈上会则样？

我猜 如果放在栈上的话，等栈内存被清空（如调用函数返回等）时这个当前值也会消失。逃逸到堆上就不会产生这个问题

**Comment by SimonHu-HN** [comment]
> @E-c-h-o-Go
> > @wuqinqiang
> > 没明白 为什么需要调用 reflect.escapes 保证当前值逃逸到堆上，如果在栈上会则样？
> 
> 我猜 如果放在栈上的话，等栈内存被清空（如调用函数返回等）时这个当前值也会消失。逃逸到堆上就不会产生这个问题

emmm，栈空间被清空，引用也就没了吧，这个值逃到堆上，也没啥意义啊，主要是不知道它为什么要逃到堆上去。我是用java思维来理解的，我才接触GO，不知道现在您对这个`escapes `的用法有没有什么新的理解，我是有点不太明白他的具体意义和使用环境的。

**Comment by liyechen** [comment]
> @SimonHu-HN
> > @E-c-h-o-Go
> > > @wuqinqiang
> > > 没明白 为什么需要调用 reflect.escapes 保证当前值逃逸到堆上，如果在栈上会则样？
> > 
> > 我猜 如果放在栈上的话，等栈内存被清空（如调用函数返回等）时这个当前值也会消失。逃逸到堆上就不会产生这个问题
> 
> emmm，栈空间被清空，引用也就没了吧，这个值逃到堆上，也没啥意义啊，主要是不知道它为什么要逃到堆上去。我是用java思维来理解的，我才接触GO，不知道现在您对这个`escapes `的用法有没有什么新的理解，我是有点不太明白他的具体意义和使用环境的。

[reflect-valueof](https://github.com/golang/go/blob/c95464f0ea3f87232b1f3937d1b37da6f335f336/src/reflect/value.go#L2701) 官方给了注释，以后可能会留在栈上，只是目前为了某某某原因，统一放到了堆上。 原因如下: [comment](https://github.com/golang/go/blob/c95464f0ea3f87232b1f3937d1b37da6f335f336/src/reflect/value.go#L3123)

**Comment by cuiweixie** [comment]
实体书， [126页](https://i.pinimg.com/originals/bb/3d/f4/bb3df41c266af94667da05f7d9b1b2d4.jpg)  
"Go 语言反射的最后一条法与值是否可以被更改有关" -> "Go 语言反射的最后一条法则是与值是否可以被更改有关"

**Comment by draveness** [comment]
> 实体书， [126页](https://i.pinimg.com/originals/bb/3d/f4/bb3df41c266af94667da05f7d9b1b2d4.jpg) "Go 语言反射的最后一条法与值是否可以被更改有关" -> "Go 语言反射的最后一条法则是与值是否可以被更改有关"

这个我记一下

**Comment by jingchu000** [comment]
大佬，这里应该是有笔误
3.如果当前函数是方法，需要向将方法的接收接收者者拷贝到 args 内存中；
修订
3.如果当前函数是方法，需要向将方法的接收者拷贝到 args 内存中；

**Comment by dgqypl** [comment]
这一节的错别字有点儿多：
这以法则 -> 这一法则
可以从反射对象~~可以~~获取
所以我们只能~~只能~~用迂回的方式
需要~~向~~将方法的~~接收~~接收者~~者~~拷贝到

**Comment by yunbyte** [comment]
reflect.Value.Call 是运行时调用方法的入口，它通过两个 MustBe 开头的方法确定了当前反射对象的类型是函数以及可见性，随后调用 reflect.Value.call 完成方法调用，这个私有方法的执行过程会分成以下的几个部分：

纸质书的 “随后调用 reflect.Value.Call 完成方法调用” 这句话中reflect.Value.call被写成了大写，在书中132页的左下角
